### PR TITLE
feat: basic support for inferrable types

### LINKED ISSUE #1406: Support inferrable types
I have the following typescript code:
```ts
export class Foo {
  a: boolean = false;
  b = false;
}
```

ts-json-schema-generator generates the following schema:
```json
{
  "$ref": "#/definitions/Foo",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "Foo": {
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "boolean"
        }
      },
      "required": [
        "a"
      ],
      "type": "object"
    }
  }
}
```

It seems like it just skips `b` because it has no explicit type. However, this is valid typescript. It is not even allowed to add types here if you use the recommended config of typescript-eslint. I don't think it's possible to get the type directly from the AST. Maybe it is possible to implement something similar to the eslint rule [no-inferrable-types](https://github.com/typescript-eslint/typescript-eslint/blob/5d0bf1ca8d972fdbf97653b841abc16da236aac9/packages/eslint-plugin/src/rules/no-inferrable-types.ts) in this project.

### PR DESCRIPTION
**Pull request by swnf**
This PR adds support to infer the type of properties without an explicit type. For example:
```ts
export class Foo {
  a = false;
}
```
Previously, properties without an explicit type were always ignored. With this PR they will get their correct type from the type checker.

This PR fixes #1406.

# Version

Published prerelease version: `v1.2.0-next.3`

<details>
  <summary>Changelog</summary>

  :tada: This release contains work from new contributors! :tada:
  
  Thanks for all your work!
  
  :heart: null[@swnf](https://github.com/swnf)
  
  :heart: Thomas ([@thomaswr](https://github.com/thomaswr))
  
  :heart: Arthur Fiorette ([@arthurfiorette](https://github.com/arthurfiorette))
  
  :heart: Sean Keenan ([@sean9keenan](https://github.com/sean9keenan))
  
  #### 🚀 Enhancement
  
  - feat: basic support for inferrable types [#1407](https://github.com/vega/ts-json-schema-generator/pull/1407) ([@swnf](https://github.com/swnf))
  - feat: add support for type URL (string with format "uri") [#1480](https://github.com/vega/ts-json-schema-generator/pull/1480) (thomas.riener@derstandard.at [@thomaswr](https://github.com/thomaswr))
  - feat: support for sourceless nodes. [#1386](https://github.com/vega/ts-json-schema-generator/pull/1386) ([@arthurfiorette](https://github.com/arthurfiorette))
  - feat: Add support for computed types on literals [#1438](https://github.com/vega/ts-json-schema-generator/pull/1438) ([@sean9keenan](https://github.com/sean9keenan))
  
  #### 🐛 Bug Fix
  
  - fix: Export HiddenType and HiddenTypeFormatter [#1437](https://github.com/vega/ts-json-schema-generator/pull/1437) ([@sean9keenan](https://github.com/sean9keenan))
  
  #### 🔩 Dependency Updates
  
  - chore(deps-dev): bump @typescript-eslint/parser from 5.43.0 to 5.44.0 [#1483](https://github.com/vega/ts-json-schema-generator/pull/1483) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump prettier from 2.7.1 to 2.8.0 [#1482](https://github.com/vega/ts-json-schema-generator/pull/1482) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.43.0 to 5.44.0 [#1484](https://github.com/vega/ts-json-schema-generator/pull/1484) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest-junit from 14.0.1 to 15.0.0 [#1479](https://github.com/vega/ts-json-schema-generator/pull/1479) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.42.1 to 5.43.0 [#1473](https://github.com/vega/ts-json-schema-generator/pull/1473) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/jest from 29.2.2 to 29.2.3 [#1474](https://github.com/vega/ts-json-schema-generator/pull/1474) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.42.1 to 5.43.0 [#1475](https://github.com/vega/ts-json-schema-generator/pull/1475) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 8.27.0 to 8.28.0 [#1476](https://github.com/vega/ts-json-schema-generator/pull/1476) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump ajv from 8.11.0 to 8.11.2 [#1478](https://github.com/vega/ts-json-schema-generator/pull/1478) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.42.0 to 5.42.1 [#1468](https://github.com/vega/ts-json-schema-generator/pull/1468) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump chai from 4.3.6 to 4.3.7 [#1469](https://github.com/vega/ts-json-schema-generator/pull/1469) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.42.0 to 5.42.1 [#1470](https://github.com/vega/ts-json-schema-generator/pull/1470) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest from 29.2.2 to 29.3.1 [#1471](https://github.com/vega/ts-json-schema-generator/pull/1471) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.41.0 to 5.42.0 [#1461](https://github.com/vega/ts-json-schema-generator/pull/1461) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-env from 7.19.4 to 7.20.2 [#1464](https://github.com/vega/ts-json-schema-generator/pull/1464) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/jest from 29.0.3 to 29.2.2 [#1462](https://github.com/vega/ts-json-schema-generator/pull/1462) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 8.26.0 to 8.27.0 [#1463](https://github.com/vega/ts-json-schema-generator/pull/1463) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.41.0 to 5.42.0 [#1465](https://github.com/vega/ts-json-schema-generator/pull/1465) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.11.7 to 18.11.9 [#1466](https://github.com/vega/ts-json-schema-generator/pull/1466) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/core from 7.19.6 to 7.20.2 [#1467](https://github.com/vega/ts-json-schema-generator/pull/1467) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.40.1 to 5.41.0 [#1458](https://github.com/vega/ts-json-schema-generator/pull/1458) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest from 29.2.1 to 29.2.2 [#1457](https://github.com/vega/ts-json-schema-generator/pull/1457) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.40.1 to 5.41.0 [#1459](https://github.com/vega/ts-json-schema-generator/pull/1459) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.11.3 to 18.11.7 [#1460](https://github.com/vega/ts-json-schema-generator/pull/1460) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps): bump safe-stable-stringify from 2.4.0 to 2.4.1 [#1448](https://github.com/vega/ts-json-schema-generator/pull/1448) ([@dependabot[bot]](https://github.com/dependabot[bot]) [@domoritz](https://github.com/domoritz))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.40.0 to 5.40.1 [#1449](https://github.com/vega/ts-json-schema-generator/pull/1449) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 8.25.0 to 8.26.0 [#1447](https://github.com/vega/ts-json-schema-generator/pull/1447) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.11.0 to 18.11.3 [#1450](https://github.com/vega/ts-json-schema-generator/pull/1450) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/core from 7.19.3 to 7.19.6 [#1451](https://github.com/vega/ts-json-schema-generator/pull/1451) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest from 29.2.0 to 29.2.1 [#1452](https://github.com/vega/ts-json-schema-generator/pull/1452) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.40.0 to 5.40.1 [#1453](https://github.com/vega/ts-json-schema-generator/pull/1453) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump vega-lite from 5.5.0 to 5.6.0 [#1444](https://github.com/vega/ts-json-schema-generator/pull/1444) ([@dependabot[bot]](https://github.com/dependabot[bot]) [@domoritz](https://github.com/domoritz))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.39.0 to 5.40.0 [#1441](https://github.com/vega/ts-json-schema-generator/pull/1441) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.8.3 to 18.11.0 [#1440](https://github.com/vega/ts-json-schema-generator/pull/1440) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest from 29.1.2 to 29.2.0 [#1442](https://github.com/vega/ts-json-schema-generator/pull/1442) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.39.0 to 5.40.0 [#1443](https://github.com/vega/ts-json-schema-generator/pull/1443) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-env from 7.19.3 to 7.19.4 [#1445](https://github.com/vega/ts-json-schema-generator/pull/1445) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.38.1 to 5.39.0 [#1435](https://github.com/vega/ts-json-schema-generator/pull/1435) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump eslint from 8.24.0 to 8.25.0 [#1432](https://github.com/vega/ts-json-schema-generator/pull/1432) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump jest from 29.0.3 to 29.1.2 [#1433](https://github.com/vega/ts-json-schema-generator/pull/1433) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.7.23 to 18.8.3 [#1434](https://github.com/vega/ts-json-schema-generator/pull/1434) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.38.1 to 5.39.0 [#1436](https://github.com/vega/ts-json-schema-generator/pull/1436) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/parser from 5.38.0 to 5.38.1 [#1426](https://github.com/vega/ts-json-schema-generator/pull/1426) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @types/node from 18.7.20 to 18.7.23 [#1424](https://github.com/vega/ts-json-schema-generator/pull/1424) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/core from 7.19.1 to 7.19.3 [#1425](https://github.com/vega/ts-json-schema-generator/pull/1425) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @babel/preset-env from 7.19.1 to 7.19.3 [#1427](https://github.com/vega/ts-json-schema-generator/pull/1427) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps-dev): bump @typescript-eslint/eslint-plugin from 5.38.0 to 5.38.1 [#1428](https://github.com/vega/ts-json-schema-generator/pull/1428) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps): bump commander from 9.4.0 to 9.4.1 [#1429](https://github.com/vega/ts-json-schema-generator/pull/1429) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  - chore(deps): bump typescript from 4.8.3 to 4.8.4 [#1430](https://github.com/vega/ts-json-schema-generator/pull/1430) ([@dependabot[bot]](https://github.com/dependabot[bot]))
  
  #### Authors: 7
  
  - [@dependabot[bot]](https://github.com/dependabot[bot])
  - [@swnf](https://github.com/swnf)
  - Arthur Fiorette ([@arthurfiorette](https://github.com/arthurfiorette))
  - Dominik Moritz ([@domoritz](https://github.com/domoritz))
  - Sean Keenan ([@sean9keenan](https://github.com/sean9keenan))
  - Thomas ([@thomaswr](https://github.com/thomaswr))
  - Thomas Riener (thomas.riener@derstandard.at)
</details>

### CODE REVIEW

File: test/valid-data/class-single/main.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export class MyObject {
     readonly readonlyProp: string;
 
     // Constructors must be ignored
-    public constructor(protected a: number, private b: number, c: number, public propC: number,
+    public constructor(protected a: number, private b: number, c: number, public propC = 42,
```
Comment (arthurfiorette) [inline test/valid-data/class-single/main.ts]: You could add a comment to help identify that the `= 42` is used to test inferrable types.

File: src/Utils/inferrableType.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    else if (
+        expression.kind === ts.SyntaxKind.FalseKeyword ||
+        expression.kind === ts.SyntaxKind.TrueKeyword ||
+        (ts.isPrefixUnaryExpression(expression) && expression.operator === ts.SyntaxKind.ExclamationToken)
+    )
+        return new BooleanType();
+    else if (expression.kind === ts.SyntaxKind.NullKeyword) return new NullType();
+    else if (expression.kind === ts.SyntaxKind.UndefinedKeyword || ts.isVoidExpression(expression))
+        return new UndefinedType();
+    else return undefined;
```
Comment (arthurfiorette) [inline src/Utils/inferrableType.ts]: This seems more readable :)

```suggestion
  if (expression == undefined) {
    return undefined;
  }

  if (ts.isParenthesizedExpression(expression)) {
    return getInferrableType(expression.expression);
  }

  if (
    ts.isNumericLiteral(expression) ||
    (ts.isPrefixUnaryExpression(expression) &&
      (expression.operator === ts.SyntaxKind.PlusPlusToken ||
        expression.operator === ts.SyntaxKind.MinusMinusToken ||
        expression.operator === ts.SyntaxKind.PlusToken ||
        expression.operator === ts.SyntaxKind.MinusToken ||
        expression.operator === ts.SyntaxKind.TildeToken))
  ) {
    return new NumberType();
  }

  if (
    ts.isStringLiteral(expression) ||
    ts.isNoSubstitutionTemplateLiteral(expression) ||
    ts.isTemplateLiteral(expression)
  ) {
    return new StringType();
  }

  if (
    expression.kind === ts.SyntaxKind.FalseKeyword ||
    expression.kind === ts.SyntaxKind.TrueKeyword ||
    (ts.isPrefixUnaryExpression(expression) && expression.operator === ts.SyntaxKind.ExclamationToken)
  ) {
    return new BooleanType();
  }

  if (expression.kind === ts.SyntaxKind.NullKeyword) {
    return new NullType();
  }

  if (expression.kind === ts.SyntaxKind.UndefinedKeyword || ts.isVoidExpression(expression)) {
    return new UndefinedType();
  }
  
  return undefined;
```

File: src/Utils/inferrableType.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import ts from "typescript";
+import { BaseType } from "../Type/BaseType";
+import { BooleanType } from "../Type/BooleanType";
+import { NullType } from "../Type/NullType";
+import { NumberType } from "../Type/NumberType";
+import { StringType } from "../Type/StringType";
+import { UndefinedType } from "../Type/UndefinedType";
+
+export function getInferrableType(expression: ts.Expression | undefined): BaseType | undefined {
+    // TODO: Ideally, this should also support function calls and property access expressions
```
Comment (arthurfiorette) [inline src/Utils/inferrableType.ts]: About function calls (a.k.a inferred response types), you'll need to use the typechecker to get the inferred type from a function. Probably like what I've done with kitajs:

https://github.com/arthurfiorette/kita/blob/48dc243d7e3f10363b33c135e758003de1912b83/packages/generator/src/util/node.ts#L41

The only problem is that sometimes functions return types are not manually typed, and when you do the above, it's generated ts.TypeNodes without a valid source. The good news is that the #1386 fixed it.

For property access expressions, I guess that it is also going to be similar to function response types.

File: src/NodeParser/InterfaceAndClassNodeParser.ts [hunk 4]
```diff
@@ -119,12 +120,20 @@ export class InterfaceAndClassNodeParser implements SubNodeParser {
                 }
                 return members;
             }, [] as (ts.PropertyDeclaration | ts.PropertySignature | ts.ParameterPropertyDeclaration)[])
-            .filter((member) => isPublic(member) && !isStatic(member) && member.type && !isNodeHidden(member))
+            .filter(
+                (member) =>
+                    isPublic(member) &&
+                    !isStatic(member) &&
+                    (member.type || getInferrableType(member.initializer)) &&
```
Comment (arthurfiorette) [inline src/NodeParser/InterfaceAndClassNodeParser.ts]: Could you do something so we don't call `getInferrableType` inside the `filter` and the `map` function?

**Review by arthurfiorette — CHANGES_REQUESTED** [review]
Hey @swnf, I'm not dedicated to this project or either a main contributor. But I arrieved at your PR randomly and decided to stick my nose for a bit, you can only use what you find useful :)

**Comment by swnf** [comment]
Thanks for your feedback @arthurfiorette. I made some changes based on it. However, I'm not so sure anymore if my approach makes sense. There does not seem to be any advantage in parsing the AST of the initializer (except maybe for an `AsExpression`). It might make more sense to just use the type checker directly to get the type of a member if it has no explicit type (would probably depend on #1386).

**Comment by domoritz** [comment]
Thanks for the pull request. I think using the type checker directly would make sense if it makes the code a lot simpler. We mainly use the AST to preserve aliases but that isn't an issue here with inferred types.

**Comment by domoritz** [comment]
@swnf would you be able to revise this pull request?

**Comment by swnf** [comment]
Yes, I think I can update it next week.

**Comment by swnf** [comment]
I've updated my code. It now uses the type checker and works for all types. I will rebase the PR once #1386 is merged.

**Comment by domoritz** [comment]
I merged https://github.com/vega/ts-json-schema-generator/pull/1386

**Comment by swnf** [comment]
I've rebased the PR. I think it is now ready for review.

**Comment by swnf** [comment]
@domoritz can you have a look at this PR? It would help me with some eslint issues.

**Pull request merged**